Spark: Using a Secret for setting AWS credentials in Spark History Server (#214)
kudobuilder/frameworks: repository/spark/docs/latest/configuration.md

If `createRBAC` flag is set to `false`, a `Role` (with a `RoleBinding` linked to Spark Service Account) should be
created or exist prior to submission of Spark Applications. `Role` configuration and a list of required permissions are
available in [spark-rbac.yaml](../../operator/templates/spark-rbac.yaml) template file.

### Integration with AWS S3
This section describes the steps required to provide an access to S3 when using Spark History server 

> or accessing S3 buckets from Spark workloads. 

Reviewer: Suggestion:

This section describes the steps required for configuring secure access to S3 buckets for Spark workloads and Spark History server.
or accessing S3 buckets from Spark workloads.

Author: done

With this method all sensitive data is stored as Kubernetes Secret and mounted to pods via environment variables.

> First, you need to convert AWS credentials to base64 as follows:

Reviewer: ```suggestion
In order to store AWS credentials in a `Secret`, the credentials should be converted to base64 first:
```

```bash
$ echo "<AWS_ACCESS_KEY_ID>" | base64                    
  QVdTX0FDQ0VTU19LRVlfSUQK
$ echo "<AWS_SECRET_ACCESS_KEY>" | base64                    
  aVJhTVlQZVM1NTJzdFNpaDhVWDNEVkRyMndaMXpZOGxtWWlOKy9TQwo=    
```
Create a `Secret` with the following contents:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: aws-credentials
  namespace: spark-operator
type: Opaque
data:
  AWS_ACCESS_KEY_ID: QVdTX0FDQ0VTU19LRVlfSUQK
  AWS_SECRET_ACCESS_KEY: aVJhTVlQZVM1NTJzdFNpaDhVWDNEVkRyMndaMXpZOGxtWWlOKy9TQwo=
```
Note: a Secret must be in the same namespace as an Spark Operator.

To read more about Secrets, refer to the [official K8s documentation](https://kubernetes.io/docs/concepts/configuration/secret/).

> Here is the example using a Secret to pass AWS credentials as environment variables to `SparkApplication`: 

Reviewer: ```suggestion
Here is an example configuration which uses a `Secret` to pass AWS credentials as environment variables to `SparkApplication`: 
```

```yaml
apiVersion: "sparkoperator.k8s.io/v1beta2"
kind: SparkApplication
metadata:
  name: spark-app
  namespace: spark-operator
spec:
  ...
  sparkConf:
    "spark.eventLog.enabled": "true"
    "spark.eventLog.dir": "s3a://<s3-location>"
    "spark.hadoop.fs.s3a.impl": "org.apache.hadoop.fs.s3a.S3AFileSystem"
  ...
  driver:
    env:
      - name: AWS_ACCESS_KEY_ID
        valueFrom:
          secretKeyRef:
            name: aws-credentials
            key: AWS_ACCESS_KEY_ID
      - name: AWS_SECRET_ACCESS_KEY
        valueFrom:
          secretKeyRef:
            name: aws-credentials
            key: AWS_SECRET_ACCESS_KEY

>       # in case when Temporary Security Credentials are used

Reviewer: 👍 really like the attention to such minor details
